tests/framework/filters/AccessControlTest.php

```diff
@@ -0,0 +1,180 @@
+
+        $this->assertInstanceOf('\Closure', $filter->denyCallback);
+    }
+
+    public function testInitSetDefault()
+    {
+        $filter = new AccessControl();
+
+        $this->assertInstanceOf('\Closure', $filter->denyCallback);
+    }
+
+    public function testBeforeActionAccessRuleAllowReturnTrue()
+    {
+        /* @var $fakeRule AccessRule|ObjectProphecy */
+        $fakeRule = $this->prophesize(AccessRule::className());
+        $fakeRule->allows(Argument::any(), Argument::any(), Argument::any())->willReturn(true);
+        Yii::$container->set(AccessRule::className(), $fakeRule->reveal());
+        $config = [
+            'rules' => [[]]
+        ];
+        $filter = new AccessControl($config);
+
+        $allow = $filter->beforeAction(null);
+
+        $this->assertTrue($allow);
+    }
+
+    public function testBeforeActionAccessRuleDenyReturnFalse()
+    {
+        /* @var $fakeRule AccessRule|ObjectProphecy */
+        $fakeRule = $this->prophesize(AccessRule::className());
+        $fakeRule->allows(Argument::any(), Argument::any(), Argument::any())->willReturn(false);
+        Yii::$container->set(AccessRule::className(), $fakeRule->reveal());
+
+        $config = [
+            'rules' => [[]],
+            'denyCallback' => function () {},
+        ];
+        $filter = new AccessControl($config);
+
+        $deny = $filter->beforeAction(null);
+
+        $this->assertFalse($deny);
+    }
+    
+    public function testBeforeActionAccessRuleDenyAndFilledDenyCallbackReturnFalse()
+    {
+        /* @var $fakeRule AccessRule|ObjectProphecy */
+        $fakeRule = $this->prophesize(AccessRule::className());
+        $isCalledRuleDenyCallback = false;
+        $fakeRule->denyCallback = function () use (&$isCalledRuleDenyCallback) {
+            $isCalledRuleDenyCallback = true;
+        };
+        $fakeRule->allows(Argument::any(), Argument::any(), Argument::any())->willReturn(false);
+        Yii::$container->set(AccessRule::className(), $fakeRule->reveal());
+        $config = [
+            'rules' => [[]],
+            'denyCallback' => function () {},
+        ];
+        $filter = new AccessControl($config);
+
+        $deny = $filter->beforeAction(null);
+
+        $this->assertFalse($deny);
+        $this->assertTrue($isCalledRuleDenyCallback);
+    }
+
+    public function testBeforeActionAccessRuleNotSuitableReturnFalse()
+    {
+        /* @var $fakeRule AccessRule|ObjectProphecy */
+        $fakeRule = $this->prophesize(AccessRule::className());
+        $fakeRule->allows(Argument::any(), Argument::any(), Argument::any())->willReturn(null);
+        Yii::$container->set(AccessRule::className(), $fakeRule->reveal());
+        $config = [
+            'rules' => [[]],
+            'denyCallback' => function () {},
+        ];
+        $filter = new AccessControl($config);
+
+        $deny = $filter->beforeAction(null);
+
+        $this->assertFalse($deny);
+    }
+
```

Comment:
the previous methods should ideally be merged into a dataProvider in which the logic is easier to understand and verified. Now the tests are bit hard to read & to compare.